### DISCUSSION TITLE
dynamically branch over targets, and get the `tar_meta()` of each branch

### DISCUSSION THREAD

**Discussion by mkoohafkan** [ropensci/targets#877] [Help]
Is it possible to pass a dynamic target's metadata to a target? I can't figure out how to dynamically branch over a target and use `tar_meta()` to get that target's metadata. See the example, I want to get the `tar_meta()` of each dynamic branch of `y`, but instead I get the `tar_meta()` of the `y` pattern.

```r
library(targets)
td = file.path(tempdir(), "test")
dir.create(td)
setwd(td)

tar_script({
  list(
    tar_target(x, seq_len(1)),
    tar_target(y, {2 * x; Sys.sleep(5)}, pattern = map(x)),
    # I want to get the metadata for each `y` branch
    # but instead I repeatedly get the metadata of `y` pattern
    tar_target(ymeta, tar_meta(y), pattern = map(y))
  )
}, ask = FALSE)
tar_make()

# what want
identical(
  tar_meta(y_29239c8a)[c("time", "seconds")],
  tar_read(ymeta_d0697b30)[c("time", "seconds")]
)

# what I get
identical(
  tar_meta(y)[c("time", "seconds")],
  tar_read(ymeta_d0697b30)[c("time", "seconds")]
)
```

My use case is that I want to write the outputs of a target to a file, and the metadata of that target to a second file (mainly the "time" and "seconds" columns of `tar_meta()`).

**Comment by wlandau** [comment]
You could get all the branch metadata using something like tar_meta(names = starts_with(“y_”)). Metadata is meant for access outside the pipeline.

**Reply by mkoohafkan** [reply]
I considered that too---but I would need to know the name of the current branch. How can I determine that?

**Reply by wlandau** [reply]
You can get the name of the branch currently running with tar_name(), but it it hard to get the right upstream branches from there. I would suggest doing things with the metadata after the pipeline is finished rather than during a target:

**Reply by mkoohafkan** [reply]
Finding the upstream branch would not be hard in my case. The showstopper is that `tar_meta()` does not include the currently running target or list it in "children" so there is no way at all to search.

```r
tar_script({
get_name = function(n, meta) {  
  list(n, meta$name)
#  upstream = meta[meta$name == n,]$depend
#  meta[meta$data == upstream,]$name

}


  list(
    tar_target(x, seq_len(1)),
    tar_target(y, {2 * x; Sys.sleep(5)}, pattern = map(x)),
    tar_target(ymeta, get_name(tar_name(), tar_meta()), pattern = map(y))
  )
}, ask = FALSE)
# it works if you run the pipeline a second time, so destroy() first 
tar_destroy(ask = FALSE)
tar_make()

tar_read(ymeta_d0697b30)
## [[1]]
## [1] "ymeta_d0697b30"
## 
## [[2]]
## [1] "get_name"   "x"          "y_29239c8a" "y"
```

I get the same result if I call `tar_meta()` from inside `get_name()` (rather than passing it as an argument).

**Reply by wlandau** [reply]
The metadata gets stored after the target finishes, another reason why I would process the metadata in bulk after the whole pipeline is done. Or you could write your own code to record runtime etc.
